src/main/java/org/kiwiproject/retry/RetryResult.java

```diff
@@ -0,0 +1,171 @@
+package org.kiwiproject.retry;
+
+import static java.util.Objects.nonNull;
+import static org.kiwiproject.collect.KiwiLists.last;
+
+import lombok.Getter;
+import org.kiwiproject.base.UUIDs;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Defines a result of one or more attempts to get some type of object. The list of errors should reflect the order
+ * in which errors occurred, e.g. the first error should be the error for the first attempt, second error should be
+ * the error for the second attempt, etc.
+ * <p>
+ * A unique identifier is automatically assigned to new {@link RetryResult} instances for easy identification, e.g.
+ * when logging retry exceptions.
+ * <p>
+ * Note that there is nothing precluding a {@link RetryResult} from representing a failure but not actually have any
+ * errors. This could happen, for example, if retry code executes a {@link java.util.function.Supplier}
+ * and that supplier is implemented to catch all exceptions and return {@code null} to force a retry. We generally
+ * recommend to implement Suppliers to throw exceptions when errors occur. They will need to be subclasses of
+ * {@link RuntimeException} since {@link java.util.function.Supplier} doesn't permit checked exceptions, plus we
+ * generally prefer unchecked exceptions.
+ *
+ * @param <T> the result type
+ */
+@Getter
+public class RetryResult<T> {
+
+    private final String resultUuid;
+    private final int numAttemptsMade;
+    private final int maxAttempts;
+    private final T object;
+    private final List<Exception> errors;
+
+    /**
+     * Create new instance.
+     *
+     * @param numAttemptsMade the number of attempts (must be less than or equal to max attempts)
+     * @param maxAttempts     the maximum number of attempts
+     * @param object          the result, or {@code null}
+     * @param errors          a non-null list containing any errors that occurred on attempts, or empty list if no
+     *                        errors occurred
+     */
+    public RetryResult(int numAttemptsMade, int maxAttempts, T object, List<Exception> errors) {
+        this.resultUuid = UUIDs.randomUUIDString();
+        this.numAttemptsMade = numAttemptsMade;
+        this.maxAttempts = maxAttempts;
+        this.object = object;
+        this.errors = errors;
```

Comment:
We should probably protect the list from being changed by making a copy.

```suggestion
        this.errors = List.copyOf(errors);
```

Reply:
Done